File: frontend/src/components/chat/providers/ohmypi/extractors/todo.test.ts
import { describe, expect, it } from 'vitest'
import { ohMyPiTodoSource } from './todo'

describe('ohMyPiTodoSource', () => {
  it('reads the snapshot omp states after an operation', () => {
    // omp 18.2.11's own details (probe s2).
    const source = ohMyPiTodoSource({ op: 'init' }, {
      op: 'init',
      phases: [{ name: 'Build', tasks: [{ content: 'Write code', status: 'in_progress' }, { content: 'Test it', status: 'pending' }] }],
      storage: 'session',
    })
    expect(source).toEqual({
      op: 'init',
      items: [
        { rowKey: '0:Write code', content: 'Write code', status: 'in_progress', activeForm: 'Write code', description: 'Build' },
        { rowKey: '1:Test it', content: 'Test it', status: 'pending', activeForm: '', description: 'Build' },
      ],
    })
  })

  it('maps every omp status and states a blocker', () => {
    const source = ohMyPiTodoSource({ op: 'view' }, {
      phases: [
        { name: 'Plan', tasks: [{ content: 'Read', status: 'completed' }, { content: 'Old', status: 'abandoned' }] },
        { name: '', tasks: [{ content: 'Deploy', status: 'blocked', blocker: 'no key' }, { content: 'Odd', status: 'hibernating' }] },
      ],
    })
    expect(source?.items.map(item => [item.content, item.status, item.description])).toEqual([
      ['Read', 'completed', 'Plan'],
      ['Old', 'deleted', 'Plan'],
      ['Deploy', 'pending', 'Blocked: no key'],
      ['Odd', 'pending', undefined],
    ])
  })

  it('skips a blank task and a malformed phase', () => {
    const source = ohMyPiTodoSource({}, { phases: [{ name: 'A', tasks: [{ content: '  ' }, 'x', { content: 'Keep' }] }, 'bad', { name: 'B' }] })
    expect(source?.items.map(item => item.content)).toEqual(['Keep'])
    expect(source?.op).toBe('')
  })

  it('keys two tasks with the same text apart', () => {
    // omp matches a task by its text, and two phases can hold the same words.
    const source = ohMyPiTodoSource({ op: 'view' }, { phases: [{ name: 'A', tasks: [{ content: 'Test it', status: 'completed' }] }, { name: 'B', tasks: [{ content: 'Test it', status: 'pending' }] }] })
    const keys = source?.items.map(item => item.rowKey) ?? []
    expect(keys).toHaveLength(2)
    expect(new Set(keys).size).toBe(2)
  })

  it('trims the text of a task and of its blocker, and states no blocker that is blank', () => {
    const source = ohMyPiTodoSource({ op: 'view' }, { phases: [{ name: 'P', tasks: [{ content: '  Deploy  ', status: 'blocked', blocker: '   ' }] }] })
    expect(source?.items).toEqual([{ rowKey: '0:Deploy', content: 'Deploy', status: 'pending', activeForm: '', description: 'P' }])
  })

  it('reads a phase whose tasks are not a list as no tasks, and a phases field that is not a list as no snapshot', () => {
    expect(ohMyPiTodoSource({}, { phases: [{ name: 'A', tasks: 'Write code' }] })).toEqual({ op: '', items: [] })
    expect(ohMyPiTodoSource({}, { phases: { name: 'A' } })).toBeNull()
  })

  it('reads an empty list as a snapshot that clears the panel', () => {
    expect(ohMyPiTodoSource({ op: 'clear' }, { phases: [] })).toEqual({ op: 'clear', items: [] })
  })

  it('answers null for a result that states no phases', () => {
    expect(ohMyPiTodoSource({}, {})).toBeNull()
    expect(ohMyPiTodoSource({}, undefined)).toBeNull()
  })
})

File: backend/internal/worker/agent/providers/ohmypi/todo_test.go
package ohmypi

import (
	"testing"

	"github.com/leapmux/leapmux/internal/worker/todoevents"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// frameTodoInit is omp 18.2.11's `todo init` end frame from probe s2.
const frameTodoInit = `{"type":"tool_execution_end","toolCallId":"call_6","toolName":"todo","result":{"content":[{"type":"text","text":"Remaining items (2)"}],"details":{"op":"init","phases":[{"name":"Build","tasks":[{"content":"Write code","status":"in_progress"},{"content":"Test it","status":"pending"}]}],"storage":"session"}},"isError":false}`

func TestExtractTodoEventReadsTheSnapshot(t *testing.T) {
	t.Parallel()
	event, ok := ompProvider{}.ExtractTodoEvent("todo", []byte(frameTodoInit), nil)
	require.True(t, ok)
	assert.Equal(t, todoevents.KindSnapshot, event.Kind)
	assert.Equal(t, []todoevents.Item{
		{ID: "Build/Write code", Content: "Write code", Description: "Build", Status: todoevents.StatusInProgress},
		{ID: "Build/Test it", Content: "Test it", Description: "Build", Status: todoevents.StatusPending},
	}, event.Snapshot)
}

func TestExtractTodoEventKeepsEveryPhaseAndStatus(t *testing.T) {
	t.Parallel()
	frame := `{"type":"tool_execution_end","toolCallId":"c","toolName":"todo","result":{"details":{"phases":[` +
		`{"name":"Plan","tasks":[{"content":"Read","status":"completed"},{"content":"Old idea","status":"abandoned"}]},` +
		`{"name":"Build","tasks":[{"content":"Deploy","status":"blocked","blocker":"waiting for the key"},{"content":"  ","status":"pending"},{"content":"Deploy","status":"pending"},{"content":"Odd","status":"hibernating"}]}` +
		`]}},"isError":false}`
	event, ok := ompProvider{}.ExtractTodoEvent("todo", []byte(frame), nil)
	require.True(t, ok)
	assert.Equal(t, []todoevents.Item{
		{ID: "Plan/Read", Content: "Read", Description: "Plan", Status: todoevents.StatusCompleted},
		{ID: "Plan/Old idea", Content: "Old idea", Description: "Plan", Status: todoevents.StatusDeleted},
		{ID: "Build/Deploy", Content: "Deploy", Description: "Build\nBlocked: waiting for the key", Status: todoevents.StatusPending},
		{ID: "Build/Deploy#2", Content: "Deploy", Description: "Build", Status: todoevents.StatusPending},
		{ID: "Build/Odd", Content: "Odd", Description: "Build", Status: todoevents.StatusPending},
	}, event.Snapshot, "a blank task is skipped, and a repeated text keeps a row of its own")
}

// A phase with no name still keeps a task's blocker, with no blank first line.
func TestExtractTodoEventKeepsTheBlockerOfAnUnnamedPhase(t *testing.T) {
	t.Parallel()
	frame := `{"type":"tool_execution_end","toolCallId":"c","toolName":"todo","result":{"details":{"phases":[` +
		`{"name":"","tasks":[{"content":" Deploy ","status":"blocked","blocker":"  waiting for the key "},{"content":"Test","status":"pending","blocker":"   "}]}` +
		`]}},"isError":false}`
	event, ok := ompProvider{}.ExtractTodoEvent("todo", []byte(frame), nil)
	require.True(t, ok)
	assert.Equal(t, []todoevents.Item{
		{ID: "/Deploy", Content: "Deploy", Description: "Blocked: waiting for the key", Status: todoevents.StatusPending},
		{ID: "/Test", Content: "Test", Description: "", Status: todoevents.StatusPending},
	}, event.Snapshot)
}

func TestExtractTodoEventClearsAnEmptyList(t *testing.T) {
	t.Parallel()
	event, ok := ompProvider{}.ExtractTodoEvent("todo", []byte(`{"type":"tool_execution_end","toolName":"todo","result":{"details":{"op":"clear","phases":[]}},"isError":false}`), nil)
	require.True(t, ok, "an empty list is a snapshot too: it clears the panel")
	assert.Empty(t, event.Snapshot)
}

func TestExtractTodoEventRefuses(t *testing.T) {
	t.Parallel()
	cases := map[string]struct {
		spanType string
		frame    string
	}{
		"another tool's span":  {spanType: "bash", frame: frameTodoInit},
		"a start frame":        {spanType: "todo", frame: `{"type":"tool_execution_start","toolCallId":"c","toolName":"todo","args":{}}`},
		"another tool's frame": {spanType: "todo", frame: `{"type":"tool_execution_end","toolName":"bash","result":{"details":{"phases":[]}},"isError":false}`},
		"a failed call":        {spanType: "todo", frame: `{"type":"tool_execution_end","toolName":"todo","result":{"details":{"phases":[]}},"isError":true}`},
		"no phases":            {spanType: "todo", frame: `{"type":"tool_execution_end","toolName":"todo","result":{"details":{}},"isError":false}`},
		"malformed JSON":       {spanType: "todo", frame: `{`},
	}
	for name, tc := range cases {
		_, ok := ompProvider{}.ExtractTodoEvent(tc.spanType, []byte(tc.frame), nil)
		assert.False(t, ok, name)
	}
}
